### PR TITLE
New Blog post  - Permanent WMI Event Handling

### PR DESCRIPTION
**Pull request by doctordns**
# PR Summary

## PR Checklist

- Do you want this post to be published on a specific date?: Y/N [Specify date]:
- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [x] Includes content related an open issue
  - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+tags: AD, WMI, WMI Eventing
+Summary: How to create a permanent event handler to detect changes in an AD Group
+---
+
+**Q:** Is there an easy way to detect and changes to important the membership of AD Groups?
+
+**A:**  Easy using PowerShell 7, WMI, and the CIM Cmdlets.
+
+## WMI
+Windows Management Instrumentation (WMI) is an important component of the Windows operating system.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:14]:
```suggestion
## WMI

Windows Management Instrumentation (WMI) is an important component of the Windows operating system.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+In most cases, you use PowerShell cmdlets, such as the SMB cmdlets, to manage your SMB shares.
+The value of WMI is that it can provide you access to more information and features that are not available using cmdlets. 
+
+In writing this article, I assume you have an understanding of WMI.
+In specific, I assume you understand WMI namespaces, classes, properties, and methods.
+If not, you might like to look at the [WMI Documentation](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page)
+And for some more details on using WMI and Powershell, look at using PowerShell 7 and WMI, look at [my recently published PowerShell 7 book](https://www.wiley.com/en-gb/PowerShell+7+for+IT+Professionals-p-9781119644705).
+In chapter 9, I devote a chapter to WMI and using the CIM cmdlets.
+To see just the scripts for that chapter, see my [GitHub repository](https://github.com/doctordns/Wiley20/tree/master/09%20-%20WMI).
+The scripts show you the basics of WMI and PowerShell 7.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:31]: One of the tenets of this blog is that we don't allow marketing. I am OK with linking to your book, in the right context.  See my suggestion at the bottom of the article. I think it reads better there and comes off like a pitch for your book.

```suggestion

```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+WMI is, in many cases, just another way to do things.
+For example, you can use WMI to create an SMB share by using the `Create` method of the **Win32_Share** class.
+For more information, see the (class documentation page for the class)[https://docs.microsoft.com/windows/win32/cimwin32prov/win32-share].
+In most cases, you use PowerShell cmdlets, such as the SMB cmdlets, to manage your SMB shares.
+The value of WMI is that it can provide you access to more information and features that are not available using cmdlets. 
+
+In writing this article, I assume you have an understanding of WMI.
+In specific, I assume you understand WMI namespaces, classes, properties, and methods.
+If not, you might like to look at the [WMI Documentation](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page)
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:27]:
```suggestion
If not, you might like to look at the [WMI Documentation](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page).
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+* Event consumer - this tells WMI which permanent event consumer to run and how to invoke the consumer, such as to run the Command Line consumer and run `Monitor.ps1`.
+* Event binding - this binds the filter (what event to look out for) to the consumer (what to do when the event occurs happens)
+
+To carry out these three operations, you inserting new objects into three specific WMI system classes.
+The WMI system class instances enable WMI to continue to process events after you stop your PowerShell session, log off, or restart your host.
+
+In the code below, you use the Command Line consumer to detect changes to the AD's Enterprise Admins group.
+Every time the change event occurs, you want WMI to run a specific script, namely `Monitor.ps1`.
+This script displays a list of the current members of the Enterprise Admins group to a log file and reports whether the membership now contains unauthorised users.
+If the script finds that an unauthorised user is now a group member, it writes details to a text file for you to review later.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:90]:
```suggestion
This script displays a list of the current members of the **Enterprise Admins** group to a log file and reports whether the membership now contains unauthorized users.
If the script finds that an unauthorized user is now a group member, it writes details to a text file for you to review later.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+## The Solution
+
+There are several steps in this solution.
+So please, fasten your seat belts, and away we go.
+
+### Setting up
+
+In this post, you want to detect whether an unauthorised user is a member of the Enterprise Admins group.
+You must first create a file of authorised users.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:100]:
```suggestion
In this post, you want to detect whether an unauthorized user is a member of the **Enterprise Admins** group.
You must first create a file of authorized users.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+Then you create two helper functions to assist you in testing the code.
+The function to delete all aspects of the WMI event filter from your host is useful unless you plan to keep the filter running forever.
+
+```powershell
+
+# 1. Create a list of valid users for the Enterprise Admins group
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:106]: Remove the leading and trailing blank lines in a code block.

```suggestion
# 1. Create a list of valid users for the Enterprise Admins group
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+    Where-Object Name -eq "EventFilter1" |
+      Remove-CimInstance
+  Get-CimInstance -Namespace ROOT\subscription CommandLineEventConsumer | 
+    Where-Object Name -eq 'EventConsumer1' |
+      Remove-CimInstance
+  Get-CimInstance -Namespace ROOT\subscription __FilterToConsumerBinding  |
+    Where-Object -FilterScript {$_.Filter.Name -eq 'EventFilter1'}   |
+      Remove-CimInstance
+}
+
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:143]:
```suggestion
}
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+```
+
+These two steps produce no output.
+When you create the ``OkUsers.txt`` file - ensure the users in the file are actually in your AD.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:147]: Single backticks for inline code.

```suggestion
When you create the `OkUsers.txt` file - ensure the users in the file are actually in your AD.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+
+### Creating the Event Consumer
+
+The next step is to create an event consumer - what you want WMI to do when it detects the event has occurred.
+In our example, you want the WMI permanent event handler COM object to run a script `Monitor.ps1` any time the event occurs.
+So whenever WMI detects a change to the Enterprise admins group, you want WMI to run the script.
+
+```powershell
+
+# 5. Create Monitor.ps1
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:197]:
```suggestion
# 5. Create Monitor.ps1
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+
+### Binding the Event Filter and the Event Consumer
+
+With the event filter and event consumer details added to WMI, you need to bind the two - telling WMI to detect THAT event and when it occurs, run THIS script.
+You could pre-create, for example, multiple event filters and event consumers.
+Once the binding is in place, WMI starts the monitoring process.
+
+```powershell
+
+# 7. Bind the filter and consumer
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:252]:
```suggestion
# 7. Bind the filter and consumer
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+  Filter   = [ref]$InstanceFilter     
+  Consumer = [ref]$InstanceConsumer
+}
+$IBHT = @{
+  Namespace = 'ROOT/subscription'
+  ClassName = '__FilterToConsumerBinding'
+  Property  = $Param
+}
+$InstanceBinding = New-CimInstance   @IBHT
+
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:263]:
```suggestion
$InstanceBinding = New-CimInstance   @IBHT
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+```
+
+### Checking your work
+
+A great way to check your work is to call the ``Get-WMIPE`` function you create earlier.
+What you should see is:
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:269]:
```suggestion
A great way to check your work is to call the `Get-WMIPE` function you create earlier.
What you should see is:
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+
+```powershell-console
+
+PS > # 8. Viewing the event registration details
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:273]:
```suggestion
PS > # 8. Viewing the event registration details
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+Name           CommandLineTemplate
+----           -------------------
+EventConsumer1 Pwsh.exe -File C:\Foo\Monitor.ps1
+
+***Bindings Defined ***
+
+Filter                                Consumer
+------                                --------
+__EventFilter (Name = "EventFilter1") CommandLineEventConsumer (Name = "EventConsumer1")
+
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:294]:
```suggestion
__EventFilter (Name = "EventFilter1") CommandLineEventConsumer (Name = "EventConsumer1")
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+So having created the event query, the event filter, the event consumer and the filter to consumer binding, you can test your work.
+The easiest way to test this is to add a new user to the group.
+Then, wait a few seconds for WMI to process the event, then look at the output.
+If everything is working correctly, you should see this output:
+
+
+```powershell-console
+
+PS > # 9. Adding a user to the Enterprise Admins group
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:307]:
```suggestion
PS > # 9. Adding a user to the Enterprise Admins group
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+
+Name          DistinguishedName
+----          -----------------
+Malcolm       CN=Malcolm,OU=IT,DC=Reskit,DC=Org
+Jerry Garcia  CN=Jerry Garcia,OU=IT,DC=Reskit,DC=Org
+Administrator CN=Administrator,CN=Users,DC=Reskit,DC=Org
+
+Unauthorized user [Malcolm] added to Enterprise Admins
+**********************************
+
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:322]:
```suggestion
**********************************
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+```
+
+### Troubleshooting
+This code, of course should "just work".
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:326]:
```suggestion
### Troubleshooting

This code, of course should "just work".
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+* Is the event class you are monitoring in the namespace you think it is in?
+* Is the `Monitor.ps1` script doing what you actually wanted
+
+You may also find that the **Microsoft-Windows-WMI-Activity/Operational** event log useful in tracking down issues.
+
+### Tidying up
+
+After you play with a WMI filter like this, make sure you clean up. 
+You probably don't want the filter to run forever, so remove it as soon as you can.
+To remove it, invoke the ``Remove-WMIPE`` function.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:338]:
```suggestion
To remove it, invoke the `Remove-WMIPE` function.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+And you should probably remove any inappropriate users from the Enterprise Admins group
+
+```powershell
+
+# 11. Tidying up
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:343]:
```suggestion
# 11. Tidying up
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+Remove-WMIPE    # invoke this function you defined above
+$RGMHT = @{
+  Identity = 'Enterprise admins'
+  Member   = 'Malcolm'
+  Confirm  = $false
+}
+Remove-ADGroupMember @RGMHT
+
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:351]:
```suggestion
Remove-ADGroupMember @RGMHT
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+  ClassName = "CommandLineEventConsumer"
+  Property  = $param
+}        
+$InstanceConsumer = New-CimInstance @ECHT
+```
+
+The monitoring script is fairly simple - each time the event occurs, it prints some information to a log file.
+Then it looks to see if the Enterprise Admins group contains unauthorised users - and if so, the script reports that fact to the log file.
+This script is fairly simple, and you can embellish. as needed.
+You could, for example, remove all unauthorised users.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:239]:
```suggestion
Then it looks to see if the Enterprise Admins group contains unauthorized users - and if so, the script reports that fact to the log file.
This script is fairly simple, and you can embellish. as needed.
You could, for example, remove all unauthorized users.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+
+This step creates no output.
+You might wish to call Get-WMIPE again to verify all three class occurrences no longer exist.
+
+
+## Summary
+
+WMI eventing is very powerful and straightforward to implement.
+There are thousands of WMI events you could subscribe to and which may help troubleshooting activities.
+In this case, you are examining unauthorised changers to an AD group.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:362]:
```suggestion
In this case, you are examining unauthorized changers to an AD group.
```

File: Posts/2021/04/tfl-WMIEventHandler.md [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+The WMI documentation does not provide a definitive guide to the events you might be interested in - at least that I can find.
```
Comment (sdwheeler) [inline Posts/2021/04/tfl-WMIEventHandler.md:363]:
```suggestion
The WMI documentation does not provide a definitive guide to the events you might be interested in - at least that I can find.
```
```suggestion
The WMI documentation does not provide a definitive guide to the events you might be interested in - at least that I can find.

For some more details on using WMI in Powershell 7, see my recently published [PowerShell 7 book](https://www.wiley.com/en-gb/PowerShell+7+for+IT+Professionals-p-9781119644705).
I devote a chapter 9 to WMI and using the CIM cmdlets.
You can find the scripts from this blog post and that chapter in my [GitHub repository](https://github.com/doctordns/Wiley20/tree/master/09%20-%20WMI).
```

**Review by sdwheeler — CHANGES_REQUESTED** [review]
See my comments and suggestions. Call me in Teams if you have questions.

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 6 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:106]: will do

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 8 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:147]: ok!

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 3 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:27]: will add the full stop.

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 4 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:90]: will revise for American English

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 5 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:100]: will update for US english

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 7 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:143]: removing the blank lines

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 9 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:197]: removing blank line

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 10 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:252]: removed blank line

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 13 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:273]: removed blank line

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 14 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:294]: removed blank line

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 16 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:322]: removed blank line

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 18 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:338]: fixed

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 19 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:343]: removed blank line.

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 20 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:351]: removed blank line

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 21 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:239]: changed to US ENglish

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 22 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:362]: fixed

File: Posts/2021/04/tfl-WMIEventHandler.md (hunk 23 above)
Comment (doctordns) [reply Posts/2021/04/tfl-WMIEventHandler.md:363]: fixed

**Comment by doctordns** [comment]
This post is being moved into a new folder since it'll be posted in MAY!

**Pull request closed without merging**